File: frontend/src/features/dynamic-simulation/algorithms/d_ucb.ts
/**
 * Discounted UCB (D-UCB) — TypeScript port
 * Reference: Garivier & Moulines (2011), arXiv:0805.3415
 */

import { BaseAlgorithm } from './base';
import { TransactionContext, ArmState, AlgorithmMetadata, HyperparameterSchema } from '../types';

export class DiscountedUCB implements BaseAlgorithm {
    private arms: string[] = [];
    private discount = 0.6;
    private nEff: Record<string, number> = {};
    private sEff: Record<string, number> = {};
    private totalNEff = 0;
    private totalSelections = 0;
    private lastScores: Record<string, number> = {};
    private lastChosen = '';

    initialize(arms: string[], config: Record<string, number>): void {
        this.arms = arms;
        this.discount = config.discount ?? 0.6;
        this.nEff = {};
        this.sEff = {};
        for (const arm of arms) {
            this.nEff[arm] = 0;
            this.sEff[arm] = 0;
        }
        this.totalNEff = 0;
        this.totalSelections = 0;
        this.lastScores = {};
        this.lastChosen = '';
    }

    select(_context: TransactionContext): string {
        this.totalSelections++;
        const scores: Record<string, number> = {};

        for (const arm of this.arms) {
            const n = this.nEff[arm];
            if (n < 1.0) {
                this.lastChosen = arm;
                this.lastScores = Object.fromEntries(
                    this.arms.map(a => [a, a === arm ? Infinity : 0])
                );
                return arm;
            }
            const sr = this.sEff[arm] / n;
            const bonus = Math.sqrt(2 * Math.log(this.totalNEff) / n);
            scores[arm] = sr + bonus;
        }

        this.lastScores = scores;
        this.lastChosen = Object.entries(scores).reduce((best, [arm, score]) =>
            score > best[1] ? [arm, score] : best, ['', -Infinity]
        )[0] as string;
        return this.lastChosen;
    }

    update(arm: string, reward: number, _context: TransactionContext): void {
        const gamma = this.discount;
        for (const a of this.arms) {
            this.nEff[a] *= gamma;
            this.sEff[a] *= gamma;
        }
        this.nEff[arm] += 1.0;
        this.sEff[arm] += reward;
        this.totalNEff = Object.values(this.nEff).reduce((s, v) => s + v, 0);
    }

    getState(): Record<string, ArmState> {
        const state: Record<string, ArmState> = {};
        for (const arm of this.arms) {
            const n = this.nEff[arm];
            state[arm] = {
                estimatedSR: n > 0 ? this.sEff[arm] / n : null,
                selectionScore: this.lastScores[arm] ?? null,
                totalSelections: Math.round(n),
                nEffective: Math.round(n * 100) / 100,
            };
        }
        return state;
    }

    explainLastDecision(): string {
        if (!this.lastChosen) return 'No decision made yet.';
        const score = this.lastScores[this.lastChosen] ?? 0;
        const n = this.nEff[this.lastChosen];
        const sr = n > 0 ? this.sEff[this.lastChosen] / n : 0;
        return `Chose '${this.lastChosen}' with D-UCB=${score.toFixed(4)} (SR=${sr.toFixed(3)}, n_eff=${n.toFixed(1)})`;
    }

    getHyperparameterSchema(): HyperparameterSchema {
        return {
            discount: {
                type: 'number',
                default: 0.6,
                min: 0.01,
                max: 1.0,
                step: 0.05,
                description: 'Discount factor (γ). Lower = forgets faster.',
            },
        };
    }

    metadata(): AlgorithmMetadata {
        return {
            name: 'Discounted UCB',
            shortName: 'D-UCB',
            description: 'Non-stationary UCB using discounted counters for gradual adaptation.',
            paper: 'Garivier & Moulines (2011)',
            paperUrl: 'https://arxiv.org/abs/0805.3415',
            category: 'bandit',
        };
    }
}

File: algorithms/d_ucb.py
"""
Discounted Upper Confidence Bound (D-UCB) algorithm.
Reference: Garivier & Moulines (2011), arXiv:0805.3415
"""

import math
from typing import List, Dict, Any
from algorithms.base import BaseAlgorithm, TransactionContext


class DiscountedUCB(BaseAlgorithm):
    """
    Non-stationary UCB using discounted counters. Recent observations
    weigh more than older ones, allowing gradual adaptation to SR drift.
    """

    def initialize(self, arms: List[str], config: Dict[str, Any]) -> None:
        self.arms = arms
        self.discount = config.get("discount", 0.6)
        self.n_eff: Dict[str, float] = {arm: 0.0 for arm in arms}
        self.s_eff: Dict[str, float] = {arm: 0.0 for arm in arms}
        self.total_n_eff = 0.0
        self.total_selections = 0
        self._last_scores: Dict[str, float] = {}
        self._last_chosen: str = ""

    def select(self, context: TransactionContext) -> str:
        self.total_selections += 1
        scores = {}

        for arm in self.arms:
            n = self.n_eff[arm]
            if n < 1.0:
                self._last_chosen = arm
                self._last_scores = {a: float('inf') if a == arm else 0.0 for a in self.arms}
                return arm

            sr = self.s_eff[arm] / n
            exploration_bonus = math.sqrt(2 * math.log(self.total_n_eff) / n)
            scores[arm] = sr + exploration_bonus

        self._last_scores = scores
        self._last_chosen = max(scores, key=scores.get)
        return self._last_chosen

    def update(self, arm: str, reward: int, context: TransactionContext) -> None:
        gamma = self.discount
        for a in self.arms:
            self.n_eff[a] *= gamma
            self.s_eff[a] *= gamma
        self.n_eff[arm] += 1.0
        self.s_eff[arm] += reward
        self.total_n_eff = sum(self.n_eff.values())

    def get_state(self) -> Dict[str, Any]:
        state = {}
        for arm in self.arms:
            n = self.n_eff[arm]
            state[arm] = {
                "estimated_sr": self.s_eff[arm] / n if n > 0 else None,
                "selection_score": self._last_scores.get(arm),
                "n_effective": round(n, 2),
                "s_effective": round(self.s_eff[arm], 2),
                "total_selections": self.total_selections,
            }
        return state

    def explain_last_decision(self) -> str:
        if not self._last_chosen or not self._last_scores:
            return "No decision made yet."
        chosen = self._last_chosen
        score = self._last_scores.get(chosen, 0)
        n = self.n_eff[chosen]
        sr = self.s_eff[chosen] / n if n > 0 else 0
        bonus = score - sr if score != float('inf') else float('inf')
        others = {a: f"{s:.4f}" for a, s in self._last_scores.items() if a != chosen}
        return (
            f"Chose '{chosen}' with D-UCB={score:.4f} "
            f"(discounted SR={sr:.3f} + bonus={bonus:.4f}, n_eff={n:.1f}). "
            f"Other scores: {others}"
        )

    def get_hyperparameter_schema(self) -> Dict[str, Any]:
        return {
            "discount": {
                "type": "number",
                "default": 0.6,
                "min": 0.01,
                "max": 1.0,
                "step": 0.05,
                "description": "Discount factor (γ). Lower = forgets faster. Applied multiplicatively to all counters at each step.",
            }
        }

    @classmethod
    def metadata(cls) -> Dict[str, str]:
        return {
            "name": "Discounted UCB",
            "short_name": "D-UCB",
            "description": "Non-stationary UCB using discounted counters. Recent observations weigh more, allowing gradual adaptation to SR drift.",
            "paper": "Garivier & Moulines (2011). On Upper-Confidence Bound Policies for Non-Stationary Bandit Problems. ALT 2011.",
            "paper_url": "https://arxiv.org/abs/0805.3415",
            "category": "bandit",
            "non_stationary": "true",
        }
